posts/build-an-event-sourcing-system-on-aws-using-dynamodb-and-cdk.md

```diff
@@ -0,0 +1,210 @@
+import * as lambda from '@aws-cdk/aws-lambda';
+import { SqsEventSource } from '@aws-cdk/aws-lambda-event-sources';
+
+export class EventSourcingDynamodbStack extends cdk.Stack {
+  constructor(scope: cdk.Construct, id: string, props?: cdk.StackProps) {
+    super(scope, id, props);
+
+    // Create queue
+    const queue = new sqs.Queue(this, 'queue', {
+      queueName: 'event-queue'
+    });
+
+    // Create lambda
+    const fn = new lambda.Function(this, 'fn', {
+      code: new lambda.AssetCode('resources'),
+      handler: 'lambda.handler',
+      runtime: lambda.Runtime.NODEJS_12_X,
+      functionName: 'event-to-dynamodb'
+    });
+
+    // Add lambda source
+    fn.addEventSource(new SqsEventSource(queue));
+
+    // Create dynamodb table
+    const sortKey = { name: 'Timestamp', type: dynamodb.AttributeType.STRING };
+    const table = new dynamodb.Table(this, 'table', {
+      partitionKey: { name: 'RequestId', type: dynamodb.AttributeType.STRING },
+      sortKey: sortKey,
+      tableName: 'event'
+    });
+
+    // Create GSI
+    table.addGlobalSecondaryIndex({
+      indexName: 'Type-Index',
+      partitionKey: { name: 'Type', type: dynamodb.AttributeType.STRING },
+      sortKey: sortKey
+    });
+    table.addGlobalSecondaryIndex({
+      indexName: 'Source-Index',
+      partitionKey: { name: 'Source', type: dynamodb.AttributeType.STRING },
+      sortKey: sortKey
+    });
+
+    // Grant write to lambda
+    table.grantWriteData(fn);
+  }
+}
+```
+
+(The code can be found on [the repository](https://github.com/Falydoor/event-sourcing-dynamodb/blob/master/lib/event-sourcing-dynamodb-stack.ts))
+
+CDK significantly simplifys your stack definition/deployment - saving you from a big headache.
+
+The commands below will deploy the whole stack:
+
+```sh
+cdk synth
+cdk deploy
+```
+
+(make sure to have your `AWS CLI` configured)
+
+## Testing
+
+In order to test the whole flow, we would need a Gateway that publishes a message to our SQS queue. For that, we will use `AWS CLI` to publish a message to the queue. The message content can be edited [here](https://github.com/Falydoor/event-sourcing-dynamodb/blob/master/test/message.json).
+
+```sh
+aws sqs send-message --queue-url ${QUEUE_URL} --message-body file://test/message.json
+```
+
+Then check in the DynamoDB table to make sure the event was created correctly:
+
+![Event](https://raw.githubusercontent.com/Falydoor/blog-usa/event-sourcing/images/2020/05/event-sourcing-event.png)
+
+If it was not - you will have to check the Lambda's logs on CloudWatch.
+
+# Conclusion
+
+Designing and deploying a whole event sourcing system on AWS is pretty straightforward using CDK and DynamoDB. By using AWS - the whole system is serverless, scalable, and enterprise ready! AWS CDK will not only save you a lot of time but also provide you more features than AWS Cloud​Formation. Also, the AWS CDK supports TypeScript, JavaScript, Python, Java, and C# allowing you to code in your preferred language.
```

Comment:
By using AWS, the whole system is serverless, scalable, and enterprise ready! _[comma instead of dash]_